sql/parser: adjust the precedence of JSONB operators

yuzefovich opened the pull request:
Previously, JSONB operators had a precedence different from the one
that Postgres uses. This commit brings it more inline with the latter
and improves the compatibility.

All of the affected operators fall under '(any other operator)'
category in
https://www.postgresql.org/docs/12/sql-syntax-lexical.html#SQL-PRECEDENCE.

Fixes: #40855.

Release justification: Category 4: Low risk, high benefit changes to
existing functionality.

Release note: None

yuzefovich committed:
sql/parser: adjust the precedence of JSONB operators

Previously, JSONB operators had a precedence different from the one
that Postgres uses. This commit brings it more inline with the latter
and improves the compatibility.

All of the affected operators fall under '(any other operator)'
category in
https://www.postgresql.org/docs/12/sql-syntax-lexical.html#SQL-PRECEDENCE.

Release justification: Category 4: Low risk, high benefit changes to
existing functionality.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/41095)
<!-- Reviewable:end -->

yuzefovich commented:
I'm not sure whether we should merge this before the 19.2 branch is cut.

rohany commented:
I don't have an opinion on merging it or not, but the changes look good to me -- it was good to follow the postgres precedence table. Can you also add tests for the other operators who's precedence has changed (or do you think that is necessary) ?

jordanlewis commented:
This feels like a pretty scary thing to merge at the last second. It'll break compatibility with some queries in theory no matter when we merge it, but at this point I'd rather wait til next release if you don't mind.

rafiss reviewed (commented):
awesome change!

Possibly, this commit should also adjust the PGJDBC expected failiures, look for `org.postgresql.test.jdbc4.JsonbTest.` in pkg/cmd/roachtest/pgjdbc_blacklist.go

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41095#-:-Lpe53wl6BlXVvrBsO8Z:b-g746zv)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis and @rafiss)



<!-- Sent from Reviewable.io -->

jordanlewis commented:
I'd be open to being convinced that this should go into 19.2 if you explain which operators would see a material change, and if there's some real customer problem we're solving wiith this.

yuzefovich commented:
I added parse tests for all operators in question (and removed the change to `~` operator). I don't see a reason to merge it now other than it would improve the compatibility. At the same time, I think it's unlikely to break anyone's code (unless someone is relying on the current difference between CockroachDB and PG), and I feel like users who ran into this problem before were likely to add extra parenthesis themselves, and those queries will not be affected.

I'll look into pgjdbc tests shortly.

yuzefovich commented:
I confirmed that this change allows for two PGJDBC tests to pass, and I updated the 19.2 blacklist (assuming that we'll merge this). AFAIU, we will not be backporting this PR onto previous releases, so the blacklists for them should remain unchanged, right?

jordanlewis commented:
Yes, I don't think we'll backport this. Upon inspection, I no longer think this is too risky to merge. I think in fact it would be nice to get this into 19.2 - it may well prevent some annoying compatibility issues for real people out there.

yuzefovich commented:
TFTRs!

bors r=jordanlewis
